light-clients/icsxx-cf-solana/src/header.rs

```diff
@@ -1,68 +1,146 @@
 use solana_sdk::{clock::Slot, hash::Hash, signature::Signature};
-use std::{collections::BTreeSet, convert::From, marker::PhantomData};
+use std::{collections::BTreeSet, convert::From, ops::Deref};
 
 #[derive(Clone, PartialEq, Eq, Debug)]
-pub struct Header<PK> {
-	pub shreds: Vec<Shred>,
-	_phantom: PhantomData<PK>,
+pub struct Header {
+	pub shreds: PreCheckedShreds,
 }
 
-impl<PK> Header<PK> {
+impl Header {
 	pub(crate) fn slot(&self) -> Slot {
 		assert!(!self.shreds.is_empty(), "Header must contain at least one shred");
 		self.shreds[0].slot()
 	}
 
-	pub fn hash(&self) -> Hash {
-		let mut shreds = self.shreds.iter().collect::<Vec<_>>();
-		shreds.sort_by_key(|s| s.index());
-		shreds.dedup_by_key(|s| s.index());
-
-		let data_shreds = shreds.iter().filter(|s| s.is_data()).cloned().collect::<Vec<_>>();
+	pub(crate) fn height(&self) -> Height {
+		Height::new(1, self.slot() as u64)
+	}
 
-		// TODO: replace with a return error
-		assert!(!data_shreds.is_empty());
+	/// Calculate the block hash of the block represented by its shreds.
+	///
+	/// The block hash is the hash of the last entry formed from data shreds.
+	/// If there are no data shreds, no last data shred in slot or it's not possible to form the
+	/// last entry an error is returned.
+	///
+	/// TODO: since only the last entry is needed to calculate hash, consider filtering out all the
+	/// other shreds
+	pub fn calculate_hash(&self) -> Result<Hash, Error> {
+		let data_shreds = self.shreds.iter().filter(|s| s.is_data()).collect::<Vec<_>>();
+		if data_shreds.is_empty() {
+			return Err(Error::NoDataShreds)
+		}
 
 		let last_data_shred = data_shreds.last().unwrap();
+		if !last_data_shred.last_in_slot() {
+			return Err(Error::LastShredNotLastInSlot)
+		}
 
-		// TODO: replace with a return error
-		assert!(last_data_shred.last_in_slot());
-		let _parent_slot = last_data_shred.parent().unwrap();
 		let completed_data_indexes = data_shreds
 			.iter()
 			.filter_map(|s| if s.data_complete() { Some(s.index()) } else { None })
 			.collect::<BTreeSet<_>>();
 		let consumed = last_data_shred.index() + 1;
 		let completed_ranges = get_completed_data_ranges(0, &completed_data_indexes, consumed);
 		let slot = self.slot();
-		let entries = get_slot_entries_in_block(slot, completed_ranges, &data_shreds).unwrap();
-		let slot_entries = entries;
-
-		let blockhash = slot_entries
-			.last()
-			.map(|entry| entry.hash)
-			.unwrap_or_else(|| panic!("Rooted slot {slot:?} must have blockhash"));
-		blockhash
+		let entries = get_slot_entries_in_block(slot, completed_ranges, &data_shreds)?;
+		let blockhash = entries.last().map(|entry| entry.hash).ok_or(Error::EntriesAreEmpty)?;
+
+		Ok(blockhash)
+	}
+}
+
+/// An immutable array of shreds that have the following properties:
+/// 1. Is non-empty
+/// 2. Is sorted
+/// 3. Doesn't contain duplicates
+/// 4. All shreds have the same slot
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct PreCheckedShreds(Vec<Shred>);
```

Comment:
I’d say `CheckedShreds` is a better name. `PreChecked` is ambiguous and might mean that those are shreds before the check, i.e. pre being checked.

Reply:
They're not fully checked (the signature wasn't checked at least), that's why the name is like so

Reply:
What do you think about `PartiallyCheckedShreds`?